### PR TITLE
Defensive Seed Proxy Reconciling

### PR DESCRIPTION
**Pull request by xrstf**
**What this PR does / why we need it**:
This makes sure that the seed-proxy controller does nothing if the required namespaces do not exist. So if there is no `monitoring-master` namespace, it will not attempt to put Grafana stuff in there.

This PR also replaces glog with Zap, uses the new predicate factory and makes use of the caching SeedGetterFactory.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: api/pkg/controller/seed-proxy/reconciler.go [hunk 1]
```diff
@@ -278,3 +302,16 @@ func (r *Reconciler) ensureMasterGrafanaProvisioning(seeds map[string]*kubermati
 
 	return nil
 }
+
+func (r *Reconciler) namespaceExists(client ctrlruntimeclient.Client, namespace string, log *zap.SugaredLogger) bool {
+	ns := corev1.Namespace{}
+	if err := client.Get(r.ctx, types.NamespacedName{Name: namespace}, &ns); err != nil {
+		if !kerrors.IsNotFound(err) {
```
Comment (alvaroaleman) [inline api/pkg/controller/seed-proxy/reconciler.go]: Wouldn't it make more sense to return an error to differentiate `namespaceDoesNotExist` from `some random error occured, retry later`?

File: api/pkg/controller/seed-proxy/reconciler.go [hunk 2]
```diff
@@ -78,114 +106,110 @@ func (r *Reconciler) garbageCollect(seeds map[string]*kubermaticv1.Seed) error {
 	}
 
 	if err := r.List(r.ctx, options, list); err != nil {
-		return fmt.Errorf("failed to list Secrets: %v", err)
+		return fmt.Errorf("failed to list secrets: %v", err)
```
Comment (alvaroaleman) [inline api/pkg/controller/seed-proxy/reconciler.go]: This one I like upercase, it makes clear that we are talking about the `kind: Secret`

File: api/pkg/controller/seed-proxy/reconciler.go (hunk 1 above)
Comment (xrstf) [reply api/pkg/controller/seed-proxy/reconciler.go]: I can't see right now how that would make the code easier. Can you help me? Wouldn't a caller then need to check for three different states (no error, notFoundErr and randomOtherErr)?

Comment (alvaroaleman) [reply api/pkg/controller/seed-proxy/reconciler.go]: Its actually not even about being easier, its about being correct.

If we encounter an error here, the caller will assume that the namespace doesn't exist, return and not do anything.

If we return the error, the callers knows there was an error, it gets propagated up and will result in the reconciliation being retried after the backoff period.

**Comment by xrstf** [comment]
@alvaroaleman you can review this again

File: api/pkg/controller/seed-proxy/reconciler.go [hunk 3]
```diff
@@ -37,39 +41,73 @@ type Reconciler struct {
 // an error if any API operation failed, otherwise will return an empty
 // dummy Result struct.
 func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	glog.V(4).Infof("Reconciling seed %q", request.NamespacedName.String())
+	log := r.log.With("seed", request.NamespacedName.String())
+	log.Debugw("reconciling seed")
```
Comment (alvaroaleman) [inline api/pkg/controller/seed-proxy/reconciler.go]: s/Debugw/Debug

File: api/pkg/controller/seed-proxy/reconciler.go [hunk 4]
```diff
@@ -37,39 +41,73 @@ type Reconciler struct {
...
+		return fmt.Errorf("failed to get seed client: %v", err)
+	}
+
+	err = client.Get(r.ctx, types.NamespacedName{Name: SeedServiceAccountNamespace}, &corev1.Namespace{})
+	if err != nil {
+		if !kerrors.IsNotFound(err) {
+			return fmt.Errorf("failed to check for namespace %s: %v", SeedServiceAccountNamespace, err)
+		}
+
+		log.Debug("skipping because seed namespace does not exist", "namespace", SeedServiceAccountNamespace)
```
Comment (alvaroaleman) [inline api/pkg/controller/seed-proxy/reconciler.go:89]: s/Debug/Debugw

File: api/pkg/controller/seed-proxy/reconciler.go [hunk 5]
```diff
@@ -95,49 +133,49 @@ func (r *Reconciler) garbageCollect(seeds map[string]*kubermaticv1.Seed) error {
...
+	log.Debug("reconciling ServiceAccounts...")
+	if err := r.reconcileSeedServiceAccounts(client, log); err != nil {
+		return fmt.Errorf("failed to ensure ServiceAccount: %v", err)
 	}
 
-	glog.V(4).Info("Reconciling roles...")
-	if err := r.ensureSeedRoles(client); err != nil {
-		return fmt.Errorf("failed to ensure role: %v", err)
+	err = client.Get(r.ctx, types.NamespacedName{Name: SeedPrometheusNamespace}, &corev1.Namespace{})
+	if err != nil && !kerrors.IsNotFound(err) {
```
Comment (alvaroaleman) [inline api/pkg/controller/seed-proxy/reconciler.go]: Maybe do one condition and add debug log for why it is being skipped if so?

**Pull request merged**